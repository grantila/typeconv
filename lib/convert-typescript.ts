import {
	convertCoreTypesToTypeScript,
	convertTypeScriptToCoreTypes,
	FromTsOptions,
	ToTsOptions,
} from 'core-types-ts'

import { Reader } from "./reader"
import { Writer } from "./writer"
import { userPackage, userPackageUrl } from "./package"


export function getTypeScriptReader( tsOptions?: FromTsOptions ): Reader
{
	return {
		kind: 'ts',
		read( source, { warn, filename } )
		{
			return convertTypeScriptToCoreTypes(
				source,
				{
					warn,
					...tsOptions,
				}
			);
		}
	};
}

export function getTypeScriptWriter( tsOptions?: ToTsOptions ): Writer
{
	return {
		kind: 'ts',
		write( doc, { warn, filename, sourceFilename } )
		{
			return convertCoreTypesToTypeScript(
				doc,
				{
					warn,
					filename,
					sourceFilename,
					userPackage,
					userPackageUrl,
					...tsOptions,
				}
			);
		}
	};
}
