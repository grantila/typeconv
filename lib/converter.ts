import {
	decorateError,
	decorateErrorMeta,
	isCoreTypesError,
	NamedType,
	NodeDocument,
	simplify,
	WarnFunction,
	ConversionResult as CoreTypesConversionResult,
} from 'core-types'

import type { Reader, ReaderOptions } from './reader'
import type { Shortcut, Writer, WriterOptions } from './writer'
import type { SourceÂ } from './file'
import { getSource, writeFile, relFile } from './file'
import { formatError } from './error'


export interface Target
{
	filename: string;
	relFilename?: string;
}

export interface ConversionResultInformation
{
	in: Omit< CoreTypesConversionResult, 'data' >;
	out: Omit< CoreTypesConversionResult, 'data' >;
}

export type ConversionResult< T > = ConversionResultInformation & { data: T };

export interface Converter
{
	/**
	 * Convert and save to file
	 */
	convert( from: Source, to: Target ): Promise< ConversionResult< void > >;

	/**
	 * Convert and return result as string
	 */
	convert( from: Source ): Promise< ConversionResult< string > >;

	/**
	 * The current working directory for this converter
	 */
	cwd: string;
}

export type ConvertMapFunction =
	( node: NamedType, index: number, array: ReadonlyArray< NamedType > ) =>
		NamedType;
export type ConvertFilterFunction =
	( node: NamedType, index: number, array: ReadonlyArray< NamedType > ) =>
		boolean;
export type ConvertTransformFunction = ( doc: NodeDocument ) => NodeDocument;

export interface ConvertOptions
{
	/**
	 * The current working directory to use when converting to or from files.
	 * Is not necessary for string-to-string conversion.
	 */
	cwd?: string;

	/**
	 * When simplify is true, the converter will let core-types _compress_ the
	 * types after having converted from {reader} format to core-types.
	 * This is usually recommended, but may cause some annotations (comments)
	 * to be dropped.
	 *
	 * @default
	 * 	true
	 */
	simplify?: boolean;

	/**
	 * Custom map function for transforming each type after it has been
	 * converted *from* the source type (and after it has been simplified),
	 * but before it's written to the target type system.
	 *
	 * If `filter` is used as well, this runs before `filter`.
	 * If `transform` is used as well, this runs before `transform`.
	 */
	map?: ConvertMapFunction;

	/**
	 * Custom filter function for filtering types after they have been
	 * converted *from* the source type.
	 *
	 * If `map` is used as well, this runs after `map`.
	 * If `transform` is used as well, this runs before `transform`.
	 */
	filter?: ConvertFilterFunction;

	/**
	 * Custom filter function for filtering types after they have been
	 * converted *from* the source type.
	 *
	 * If `map` is used as well, this runs after `map`.
	 * If `filter` is used as well, this runs after `filter`.
	 */
	transform?: ConvertTransformFunction;

	/**
	 * Shortcut reader and writer if possible (bypassing core-types).
	 *
	 * @default
	 *  true
	 */
	shortcut?: boolean;
}

export function makeConverter(
	reader: Reader,
	writer: Writer,
	options?: ConvertOptions
)
: Converter
{
	const { shortcut = true, cwd = process.cwd( ) } = options ?? { };

	const relFilename = ( filename: string ) =>
		relFile( options?.cwd, filename );

	async function convert( from: Source, to?: Target )
	: Promise< ConversionResult< void | string > >
	{
		const { data, filename } = await getSource( from );

		const warn: WarnFunction = ( msg, meta ) =>
		{
			const fullMeta = decorateErrorMeta(
				{ ...meta },
				{ filename, source: data }
			);
			console.warn( formatError( msg, fullMeta ) );
		}

		const toFilename =
			to?.relFilename ?? relFilename( to?.filename ?? '' );

		const readOpts: ReaderOptions = {
			warn,
			filename: filename ? relFilename( filename ) : undefined,
		};

		const writeOpts: WriterOptions = {
			warn,
			...(
				to
				? { filename: toFilename }
				: { }
			),
			...(
				filename
				? { sourceFilename: relFilename( filename ) }
				: { }
			),
			rawInput: data,
		};

		const convertShortcut = async ( ) =>
		{
			const convertedTypes: Array< string > = [ ];
			const notConvertedTypes: Array< string > = [ ];

			const shortcut = writer.shortcut?.[ reader.kind ] as Shortcut;
			const {
				data: output,
				convertedTypes: outConvertedTypes,
				notConvertedTypes: outNotConvertedTypes,
			} = await shortcut( data, readOpts, writeOpts, reader );

			return {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			};
		};

		const convertDefault = async ( ) =>
		{
			const { data: doc, convertedTypes, notConvertedTypes } =
				await reader.read( data, readOpts );

			const simplifiedDoc =
				options?.simplify === false
				? doc
				: simplify( doc );

			const { map, filter, transform } = options ?? { };

			const mappedDoc =
				typeof map === 'function'
				? {
					...simplifiedDoc,
					types: doc.types.map( ( type, index ) =>
						map( type, index, doc.types )
					),
				}
				: simplifiedDoc;

			const filteredDoc =
				typeof filter === 'function'
				? {
					...mappedDoc,
					types: doc.types.filter( ( type, index ) =>
						filter( type, index, doc.types )
					),
				}
				: mappedDoc;

			const finalDoc =
				typeof transform === 'function'
				? transform( filteredDoc )
				: filteredDoc;

			const {
				data: output,
				convertedTypes: outConvertedTypes,
				notConvertedTypes: outNotConvertedTypes
			} = await writer.write( finalDoc, writeOpts );

			return {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			};
		};

		try
		{
			const {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			} =
				( shortcut && writer.shortcut?.[ reader.kind ] )
				? await convertShortcut( )
				: await convertDefault( );

			const info: ConversionResultInformation = {
				in: { convertedTypes, notConvertedTypes },
				out: {
					convertedTypes: outConvertedTypes,
					notConvertedTypes: outNotConvertedTypes,
				},
			}

			if ( typeof to?.filename === 'undefined' )
				return { data: output, ...info } as ConversionResult< string >;

			// Only write non-empty files
			if ( outConvertedTypes.length > 0 )
				await writeFile( to?.filename, output );

			return info as ConversionResult< void >;
		}
		catch ( err )
		{
			if ( isCoreTypesError( err ) )
				decorateError( err, {
					source: data,
					...( filename ? { filename } : { } ),
				} );
			throw err;
		}
	}

	return { convert: convert as Converter[ 'convert' ], cwd };
}
