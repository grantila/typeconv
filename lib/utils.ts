
export function stringify( value: any )
{
	return JSON.stringify( value, null, 2 );
}
