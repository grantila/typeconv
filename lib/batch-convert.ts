import * as path from 'path'

import { map } from "already"

import { Converter, Target } from "./converter"
import { glob, reRootFiles } from './file'


export interface BatchConvertOptions
{
	outputDirectory?: string;
	outputExtension: string;
	verbose?: boolean;
	dryRun?: boolean;
	concurrency?: number;
}

export interface BatchConvertGlobOptions extends BatchConvertOptions
{
	hidden?: boolean;
	filesTransform?: ( files: Array< string > ) => Array< string >;
}

export interface BatchConvertResult
{
	files: number;
	types: number;
}

export async function batchConvert(
	converter: Converter,
	filenames: Array< string >,
	options: BatchConvertOptions
)
: Promise< BatchConvertResult >
{
	const {
		outputExtension,
		verbose,
		dryRun,
		// Not really CPU concurrency in this case, but enough I/O concurrency
		// to always have files prepared/read/written for CPU-bound conversion.
		// Set to 1 for deterministic unit tests.
		concurrency = 16
	} = options;

	const { cwd, convert } = converter;

	const { root, newRoot, files } =
		reRootFiles( filenames, cwd, options.outputDirectory );

	if ( verbose )
	{
		console.error( `Converting files relative to ${root}` );
		if ( newRoot !== root )
			console.error( `Storing files in ${newRoot}` );
	}

	let convertedTypes = 0;

	await map(
		files,
		{ concurrency },
		async ( { in: filename, out: outFilename, rel } ) =>
		{
			const to: Target =
				( dryRun || outputExtension === '-' )
				? undefined as unknown as Target
				: {
					filename: changeExtension( outFilename, outputExtension ),
					relFilename: changeExtension( rel, outputExtension ),
				};

			const { data, ...info } = await convert( { filename, cwd }, to );

			convertedTypes += info.out.convertedTypes.length;

			if ( verbose )
			{
				const outputRel = changeExtension( rel, outputExtension );
				const outName = outputExtension === '-' ? 'stdout' : outputRel;

				const allInputTypes =
					info.in.convertedTypes.length +
					info.in.notConvertedTypes.length;
				const notConverted =
					info.in.notConvertedTypes.length +
					info.out.notConvertedTypes.length;
				const converted = info.out.convertedTypes.length;
				const percent =
					allInputTypes === 0
					? 'no'
					: `${Math.round( converted * 100 / allInputTypes )}%`;

				console.error(
					`Converted ${rel} -> ${outName}, ${percent} types ` +
					`converted (${converted}/${allInputTypes})` +
					( !notConverted ? '' : `, ${notConverted} rejected` )
				);
			}

			if ( outputExtension === '-' && !dryRun )
				console.log( data );
		}
	);

	return {
		types: convertedTypes,
		files: files.length,
	}
}

export async function batchConvertGlob(
	converter: Converter,
	globs: Array< string >,
	options: BatchConvertGlobOptions
)
: Promise< BatchConvertResult >
{
	const { hidden, filesTransform = v => v } = options;
	const { cwd } = converter;
	const files = await glob( globs, cwd, hidden );
	return batchConvert( converter, filesTransform( files ), options );
}

function changeExtension( filename: string, extension: string )
{
	extension = extension.startsWith( '.' ) ? extension : `.${extension}`;

	return path.join(
		path.dirname( filename ),
		path.basename( filename, path.extname( filename ) ) + extension
	);
}
