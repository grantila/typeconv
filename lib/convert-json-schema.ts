import {
	convertJsonSchemaToCoreTypes,
	convertCoreTypesToJsonSchema,
	convertOpenApiToCoreTypes,
	convertCoreTypesToOpenApi,
	jsonSchemaDocumentToOpenApi,
	openApiToJsonSchema,
} from "core-types-json-schema"
import {
	JsonSchemaDocumentToOpenApiOptions,
	PartialOpenApiSchema,
} from "openapi-json-schema"
import { load as readYaml, dump as writeYaml } from "js-yaml"
import * as path from "path"

import { stringify } from "./utils"
import { Reader, ReaderOptions } from "./reader"
import { Writer } from "./writer"
import { userPackage, userPackageUrl } from "./package"


function maybeYamlReader( data: string, { warn, filename }: ReaderOptions )
: string | PartialOpenApiSchema
{
	const file = filename?.toLowerCase( ) ?? '';
	const isYaml = file.endsWith( 'yml' ) || file.endsWith( 'yaml' );

	const input = isYaml
		? readYaml(
			data,
			{
				filename,
				onWarning( { message, stack } )
				{
					warn( message, { blob: { stack } } );
				},
			}
		) as PartialOpenApiSchema
		// TODO: Maybe test if JSON parsing fails, and fallback to yaml,
		//       despite no filename match
		: data;

	return input;
}

export function getJsonSchemaReader( ): Reader
{
	return {
		kind: 'jsc',
		read( schema )
		{
			return convertJsonSchemaToCoreTypes( schema );
		},
	};
}

export function getJsonSchemaWriter( ): Writer
{
	return {
		kind: 'jsc',
		write( doc, { filename, sourceFilename } )
		{
			const { data: schemaObject, ...info } =
				convertCoreTypesToJsonSchema(
					doc,
					{ filename, sourceFilename, userPackage, userPackageUrl }
				);
			return {
				data: stringify( schemaObject ),
				...info,
			};
		},
		shortcut: {
			oapi( schema, readerOptions )
			{
				const openApiSchema = maybeYamlReader( schema, readerOptions );
				const jsonSchema = openApiToJsonSchema( openApiSchema );
				return {
					data: stringify( jsonSchema ),
					convertedTypes:
						Object.keys( jsonSchema.definitions ?? { } ),
					notConvertedTypes: [ ],
				};
			}
		},
	};
}

export function getOpenApiReader( ): Reader
{
	return {
		kind: 'oapi',
		read( schema, readerOptions )
		{
			const openApiSchema = maybeYamlReader( schema, readerOptions );
			return convertOpenApiToCoreTypes( openApiSchema );
		},
	};
}

export interface OpenAPIWriterOptions
	extends JsonSchemaDocumentToOpenApiOptions
{
	format: string; // 'yaml' | 'json'
}

export function getOpenApiWriter(
	{ format, ...openApiOptions }: OpenAPIWriterOptions
)
: Writer
{
	const formatOutput = ( data: unknown ) =>
		( format === 'yaml' || format === 'yml' )
		? writeYaml( data )
		: stringify( data );

	const getOpenApiOptions = ( filename?: string ) => ( {
		...openApiOptions,
		title:
			openApiOptions.title ??
			(
				filename
				? `Converted from ${path.basename( filename )} with typeconv`
				: 'Converted with typeconv'
			),
		version: openApiOptions.version ?? '1',
	} );

	return {
		kind: 'oapi',
		write( doc, { filename, sourceFilename } )
		{
			const { data: schemaObject, ...info } =
				convertCoreTypesToOpenApi(
						doc,
						{
							filename,
							sourceFilename,
							userPackage,
							userPackageUrl,
							...getOpenApiOptions( filename ),
						}
					);
			return {
				data: formatOutput( schemaObject ),
				...info,
			};
		},
		shortcut: {
			jsc( data, { filename }, writeOptions )
			{
				const jsonSchema = JSON.parse( data );
				const openApiSchemaObject =
					jsonSchemaDocumentToOpenApi(
						jsonSchema,
						getOpenApiOptions( filename )
					);
				return {
					data: formatOutput( openApiSchemaObject ),
					convertedTypes:
						Object.keys( jsonSchema.definitions ?? { } ),
					notConvertedTypes: [ ],
				};
			}
		},
	};
}
