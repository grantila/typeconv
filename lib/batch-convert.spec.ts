import * as path from 'path'

import { batchConvert, batchConvertGlob } from './batch-convert'
import { Converter, makeConverter } from './converter'
import { withConsoleLog } from '../test/utils'
import { getTypeScriptReader } from './convert-typescript'
import { getJsonSchemaWriter } from './convert-json-schema'


const fixtureDir = path.join( __dirname, '..', 'fixtures' );

function makeMockConverter( ): Converter
{
	const reader = getTypeScriptReader( );
	const writer = getJsonSchemaWriter( );
	return makeConverter( reader, writer, { cwd: fixtureDir } );
}

describe( "batch-convert", ( ) =>
{
	it( "batchConvert", withConsoleLog( async ( logMock ) =>
	{
		const converter = makeMockConverter( );
		await batchConvert(
			converter,
			[
				path.join( fixtureDir, 'types1.ts' ),
				path.join( fixtureDir, 'sub', 'types2.ts' ),
			],
			{
				outputExtension: '-',
				concurrency: 1,
			}
		);

		expect( logMock.mock.calls.length ).toBe( 2 );
		const types1 = JSON.parse( logMock.mock.calls[ 0 ] );
		const types2 = JSON.parse( logMock.mock.calls[ 1 ] );

		expect( types1 ).toMatchSnapshot( );
		expect( types2 ).toMatchSnapshot( );
	} ) );

	it( "batchConvertGlob", withConsoleLog( async ( logMock ) =>
	{
		const converter = makeMockConverter( );
		await batchConvertGlob(
			converter,
			[ path.join( fixtureDir, '**/*.ts' ) ],
			{
				outputExtension: '-',
				concurrency: 1,
				filesTransform: files => files.sort( ), // be deterministic
			}
		);

		expect( logMock.mock.calls.length ).toBe( 2 );
		const types1 = JSON.parse( logMock.mock.calls[ 0 ] );
		const types2 = JSON.parse( logMock.mock.calls[ 1 ] );

		expect( types1 ).toMatchSnapshot( );
		expect( types2 ).toMatchSnapshot( );
	} ) );
} );
