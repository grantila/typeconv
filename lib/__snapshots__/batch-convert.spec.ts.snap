// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batch-convert batchConvert 1`] = `
Object {
  "$comment": "Generated from types1.ts by core-types-json-schema (https://github.com/grantila/core-types-json-schema) on behalf of typeconv (https://github.com/grantila/typeconv)",
  "definitions": Object {
    "Foo": Object {
      "additionalProperties": false,
      "properties": Object {
        "bar": Object {
          "type": "string",
        },
        "baz": Object {
          "anyOf": Array [
            Object {
              "type": "number",
            },
            Object {
              "$ref": "#/definitions/Foo",
            },
          ],
          "description": "@see baz documentation at http://documentation.yada",
          "title": "A baz is a number of another Foo",
        },
      },
      "required": Array [
        "bar",
        "baz",
      ],
      "type": "object",
    },
    "Foo2": Object {
      "additionalProperties": false,
      "properties": Object {
        "bak": Object {
          "items": Object {
            "type": "boolean",
          },
          "type": "array",
        },
      },
      "required": Array [
        "bak",
      ],
      "type": "object",
    },
  },
}
`;

exports[`batch-convert batchConvert 2`] = `
Object {
  "$comment": "Generated from sub/types2.ts by core-types-json-schema (https://github.com/grantila/core-types-json-schema) on behalf of typeconv (https://github.com/grantila/typeconv)",
  "definitions": Object {
    "SubDirType": Object {
      "additionalProperties": false,
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
  },
}
`;

exports[`batch-convert batchConvertGlob 1`] = `
Object {
  "$comment": "Generated from sub/types2.ts by core-types-json-schema (https://github.com/grantila/core-types-json-schema) on behalf of typeconv (https://github.com/grantila/typeconv)",
  "definitions": Object {
    "SubDirType": Object {
      "additionalProperties": false,
      "properties": Object {
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
  },
}
`;

exports[`batch-convert batchConvertGlob 2`] = `
Object {
  "$comment": "Generated from types1.ts by core-types-json-schema (https://github.com/grantila/core-types-json-schema) on behalf of typeconv (https://github.com/grantila/typeconv)",
  "definitions": Object {
    "Foo": Object {
      "additionalProperties": false,
      "properties": Object {
        "bar": Object {
          "type": "string",
        },
        "baz": Object {
          "anyOf": Array [
            Object {
              "type": "number",
            },
            Object {
              "$ref": "#/definitions/Foo",
            },
          ],
          "description": "@see baz documentation at http://documentation.yada",
          "title": "A baz is a number of another Foo",
        },
      },
      "required": Array [
        "bar",
        "baz",
      ],
      "type": "object",
    },
    "Foo2": Object {
      "additionalProperties": false,
      "properties": Object {
        "bak": Object {
          "items": Object {
            "type": "boolean",
          },
          "type": "array",
        },
      },
      "required": Array [
        "bak",
      ],
      "type": "object",
    },
  },
}
`;
